python/cuml/internals/array.py

```diff
@@ -0,0 +1,1157 @@
+        allocation is tied. If provided, a reference to this
+        object is kept in this Buffer.
+    dtype : data-type, optional
+        Any object that can be interpreted as a numpy or cupy data type.
+    shape : int or tuple of ints, optional
+        Shape of created array.
+    order: string, optional
+        Whether to create a F-major or C-major array.
+    mem_type: {'host', 'device'}, optional
+        Whether data are on host or device.
+    validate: bool, default=None
+        Whether or not to check final array attributes against input options.
+        If None, validation will occur only for CumlArray input and input that
+        does not implement the array interface protocol and for which
+        additional options were explicitly specified.
+
+    Attributes
+    ----------
+
+    ptr : int
+        Pointer to the data
+    size : int
+        Size of the array data in bytes
+    _owner : Python Object
+        Object that owns the data of the array
+    shape : tuple of ints
+        Shape of the array
+    order : {'F', 'C'}
+        'F' or 'C' to indicate Fortran-major or C-major order of the array
+    strides : tuple of ints
+        Strides of the data
+    mem_type : MemoryType
+        Memory type for how data are stored
+    __array_interface__ : dictionary
+        ``__array_interface__`` to interop with other libraries. This
+        attribute is only present if data are host-accessible.
+    __cuda_array_interface__ : dictionary
+        ``__cuda_array_interface__`` to interop with other libraries. This
+        attribute is only present if data are device-accessible.
+
+    Notes
+    -----
+
+    cuml Array is not meant as an end-user array library. It is meant for
+    cuML/RAPIDS developer consumption. Therefore it contains the minimum
+    functionality. Its functionality is hidden by base.pyx to provide
+    automatic output format conversion so that the users see the important
+    attributes in whatever format they prefer.
+
+    Todo: support cuda streams in the constructor. See:
+    https://github.com/rapidsai/cuml/issues/1712
+    https://github.com/rapidsai/cuml/pull/1396
+
+    """
+
+    @nvtx_annotate(message="internals.CumlArray.__init__", category="utils",
+                   domain="cuml_python")
+    def __init__(self,
+                 data=None,
+                 index=None,
+                 owner=None,
+                 dtype=None,
+                 shape=None,
+                 order=None,
+                 strides=None,
+                 mem_type=None,
+                 validate=None):
+
+        if dtype is not None:
+            dtype = GlobalSettings().xpy.dtype(dtype)
+
+        self._index = index
+        if mem_type is not None:
+            mem_type = MemoryType.from_str(mem_type)
+        self._mem_type = mem_type
+
+        # Coerce data into an array interface and determine mem_type and owner
+        # if necessary
+        try:
```

Comment:
The solution for readability is more modularization, which we've scheduled for a follow-on. As for the basic question of using try/excepts for this, it's more [idiomatic](https://docs.python.org/3.4/glossary.html#term-eafp), more performant, and allows us to do more duck typing. `AttributeError` is precisely the problem we're worried about, so that's what we catch. Anything else should (correctly) halt execution.